### PR TITLE
Change remote inference to run on Cloud TPU

### PR DESCRIPTION
**Pull request by tommadams**

### CODE REVIEW

File: cc/main.cc [hunk 1]
```diff
@@ -102,8 +95,39 @@ DEFINE_double(
...
+              "When running 'eval' mode, provide a path to a second minigo "
+              "model, also serialized as a GraphDef proto.");
+DEFINE_bool(remote_inference, false,
+            "If true, run the model using the InferenceServer. This launches "
+            "a Python subprocess that performs the actual inference. Required "
+            "when running on Cloud TPU.");
+DEFINE_int32(port, 50051,
+             "The port opened by the InferenceService server.");
+DEFINE_string(tpu_name, "", "Cloud TPU name, e.g. grpc://10.240.2.2:8470.");
+DEFINE_int32(parallel_games, 32, "Number of games to play in parallel.");
```
Comment (amj) [inline cc/main.cc]: is there a comment somewhere that parallel_games should equal games_per_inference * 2 for perf?

File: cc/main.cc [hunk 2]
```diff
@@ -118,14 +142,109 @@ DEFINE_string(mode, "", "Mode to run in: \"selfplay\", \"eval\" or \"gtp\"");
...
+  }
+
+  virtual ~PlayerFactory() {
+    inference_worker_thread_.join();
+  }
+
+  virtual std::unique_ptr<MctsPlayer> New(const MctsPlayer::Options& options) = 0;
+
+  float rnd() {
+    absl::MutexLock lock(&mutex_);
```
Comment (amj) [inline cc/main.cc]: is this lock on rnd still needed?

File: cc/mcts_player.cc [hunk 3]
```diff
@@ -134,11 +137,13 @@ Coord MctsPlayer::SuggestMove() {
...
-            << absl::ToInt64Milliseconds(elapsed) << "ms"
-            << " over " << num_readouts
-            << " readouts (batched: " << options_.batch_size << ")"
-            << std::endl;
+  if (options_.verbose) {
+    std::cerr << "Milliseconds per 100 reads: "
+              << absl::ToInt64Milliseconds(elapsed) << "ms"
+              << " over " << num_readouts
+              << " readouts (batched: " << options_.batch_size << ")"
+              << std::endl;
```
Comment (amj) [inline cc/mcts_player.cc]: wanna add "estimated global ms per inference"?

File: inference_worker.py [hunk 4]
```diff
@@ -34,177 +36,159 @@
...
-        except Exception:  # pylint: disable=broad-except
+        except grpc.RpcError:
             print("Waiting for server")
             time.sleep(1)
 
 
-def wrapped_model_inference_fn(config):
-    """Wraps dual_net.model_inference_fn in a loop & RPC ops.
+def const_model_inference_fn(features):
+    def custom_getter(getter, name, *args, **kwargs):
```
Comment (amj) [inline inference_worker.py]: probably worth a comment

File: cc/dual_net/inference_server.cc [hunk 5]
```diff
@@ -34,140 +34,148 @@ using grpc::ServerContext;
...
 namespace minigo {
 
-namespace {
+namespace internal {
 
 // Implementation of the InferenceService.
-// The client InferenceServer pushes inference requests onto ServiceImpl's
-// request queue.
-class ServiceImpl final : public InferenceService::Service {
+// The client InferenceServer pushes inference requests onto
```
[the remark this answers is not in the document]
Comment (tommadams) [reply cc/dual_net/inference_server.cc]: done

File: cc/main.cc (hunk 1 above)
Comment (tommadams) [reply cc/main.cc]: done.

File: cc/main.cc (hunk 2 above)
Comment (tommadams) [reply cc/main.cc]: yes. added a comment

File: cc/mcts_player.cc (hunk 3 above)
Comment (tommadams) [reply cc/mcts_player.cc]: I'm going to do that in a different PR when I move the logging over read performance into the DualNet.

File: inference_worker.py (hunk 4 above)
Comment (tommadams) [reply inference_worker.py]: done.

**Pull request merged**